tests/test_image.py

```diff
@@ -208,6 +212,75 @@ def test_save_as_foo(self):
         self.assertFalse(image.save_as_jpeg.mock_calls)
 
 
+@mock.patch("willow.optimizers.base.OptimizerBase.process")
+class TestOptimizeImage(unittest.TestCase):
+    class DummyOptimizer(OptimizerBase):
+        library_name = "dummy"
+        image_format = "jpeg"
+
+        @classmethod
+        def check_library(cls) -> bool:
+            return True
+
+    def setUp(self):
+        with mock.patch.dict(os.environ, {"WILLOW_OPTIMIZERS": "true"}):
+            registry.register_optimizer(self.DummyOptimizer)
+
+        self.image = Image()
+
+    def tearDown(self):
+        # reset the registry as we get the global state
+        registry._registered_optimizers = []
+
+    def test_optimize_with_file_path(self, mock_process):
+        self.image.optimize("outfile", "jpeg")
+        mock_process.assert_called_with("outfile")
+
+    @mock.patch("willow.image.NamedTemporaryFile")
+    @mock.patch("willow.image.os.unlink")
+    def test_optimize_with_bytes(
+        self, mock_unlink, mock_named_temporary_file, mock_process
+    ):
+        mock_named_temporary_file.return_value.__enter__.return_value.name = "tempfile"
+        self.image.optimize(b"outfile", "jpeg")
+        mock_process.assert_called_with("tempfile")
+        mock_unlink.assert_called_with("tempfile")
+
+    @mock.patch("willow.image.NamedTemporaryFile")
+    @mock.patch("willow.image.os.unlink")
+    @mock.patch("builtins.open", mock.mock_open(read_data=b"test"))
+    def test_optimize_with_spooled_temporary_file(
+        self, mock_unlink, mock_named_temporary_file, mock_process
+    ):
```

Comment:
The argument order here appears not to match the order of the patches.

Reply:
That is correct. You can think of 

```python
@a
@b
def func():
    ...
```

as `a(b(func())` so b passes its param first, then a. in our case unlink, then named temp file

https://stackoverflow.com/a/27342171